File: src/main/java/xyz/phanta/cbtweaker/integration/jei/ingredient/JeiFluidIngredient.java
package xyz.phanta.cbtweaker.integration.jei.ingredient;

import io.github.phantamanta44.libnine.util.render.FluidRenderUtils;
import mezz.jei.api.ingredients.VanillaTypes;
import mezz.jei.api.recipe.IIngredientType;
import net.minecraft.client.util.ITooltipFlag;
import net.minecraft.util.text.TextFormatting;
import net.minecraftforge.fluids.FluidStack;

import javax.annotation.Nullable;
import java.util.Collections;
import java.util.List;

public class JeiFluidIngredient implements JeiIngredient<FluidStack> {

    private final FluidStack fluid;
    private final Role role;

    public JeiFluidIngredient(FluidStack fluid, Role role) {
        this.fluid = fluid;
        this.role = role;
    }

    public FluidStack getFluid() {
        return fluid;
    }

    @Override
    public List<FluidStack> getIngredients() {
        return Collections.singletonList(fluid);
    }

    @Nullable
    @Override
    public IIngredientType<FluidStack> getJeiIngredientType() {
        return VanillaTypes.FLUID;
    }

    @Override
    public Role getRole() {
        return role;
    }

    @Override
    public void renderIcon(int x, int y, FluidStack ingredient) {
        FluidRenderUtils.renderFluidIntoGuiCleanly(x, y, 16, 16, ingredient, ingredient.amount);
    }

    @Override
    public void getTooltip(FluidStack ingredient, List<String> tooltip, ITooltipFlag tooltipFlags) {
        tooltip.add(ingredient.getFluid().getRarity(ingredient).color + ingredient.getLocalizedName());
        tooltip.add(TextFormatting.GRAY + String.format("%,d mB", ingredient.amount));
    }

}

File: src/main/kotlin/st/evening/mc/cbtweaker/compat/jei/ingredient/impl/JeiFluidIngredient.kt
package st.evening.mc.cbtweaker.compat.jei.ingredient.impl

import mezz.jei.api.ingredients.VanillaTypes
import mezz.jei.api.recipe.IIngredientType
import net.minecraft.client.util.ITooltipFlag
import net.minecraft.util.text.TextFormatting
import net.minecraftforge.fluids.FluidStack
import st.evening.mc.cbtweaker.compat.jei.ingredient.JeiIngredient
import st.evening.mc.prelude.api.util.game.ClientSide
import st.evening.mc.prelude.api.util.game.getStillSprite
import st.evening.mc.prelude.api.util.render.RenderingHelper
import st.evening.mc.prelude.api.util.render.TextureResource
import st.evening.mc.prelude.api.util.render.gui.GuiRenderHelper

class JeiFluidIngredient(
    val fluidStack: FluidStack,
    val unitName: String,
    override val role: JeiIngredient.Role,
    override val annotation: JeiIngredient.Annotation?
) : JeiIngredient<FluidStack> {
    constructor(
        fluidStack: FluidStack,
        isRate: Boolean,
        role: JeiIngredient.Role,
        annotation: JeiIngredient.Annotation?
    ) : this(fluidStack, if (isRate) "mB/t" else "mB", role, annotation)

    override val jeiIngredientType: IIngredientType<FluidStack>
        get() = VanillaTypes.FLUID

    override fun getIngredients(): List<FluidStack> = listOf(fluidStack)

    @ClientSide.Physical
    override fun drawIcon(x: Int, y: Int, ingredient: FluidStack, partialTicks: Float) {
        TextureResource.ITEM_BLOCK_ATLAS.bind()
        RenderingHelper.setColourArgb(ingredient.fluid.getColor(ingredient))
        GuiRenderHelper.drawAtlasSprite(x, y, x + 16, y + 16, ingredient.getStillSprite())
        RenderingHelper.resetColour()
        annotation?.drawAnnotation(x + 16, y, partialTicks)
    }

    @ClientSide.Physical
    override fun getTooltip(ingredient: FluidStack, tooltip: MutableList<String>, tooltipFlags: ITooltipFlag) {
        tooltip += "${ingredient.fluid.getRarity(ingredient).color}${ingredient.localizedName}"
        tooltip += "${TextFormatting.GRAY}%,d %s".format(ingredient.amount, unitName)
        annotation?.getAnnotationTooltip(tooltip, tooltipFlags)
    }
}
